oggfilter.1
.\"
.\"  "THE BEER-WARE LICENSE" (Revision 42):
.\"  <tobias.rehbein@web.de> wrote this file. As long as you retain this notice
.\"  you can do whatever you want with this stuff. If we meet some day, and you
.\"  think this stuff is worth it, you can buy me a beer in return.
.\"                                                               Tobias Rehbein
.\"  
.
.Dd March 25, 2009
.Os 
.Dt OGGFILTER 1
.
.Sh NAME
.
.Nm oggfilter
.Nd filter a list of ogg/vorbis files using various criteria
.
.
.Sh SYNOPSIS
.
.Nm
.Op Fl l | Fl -min-length Ar period
.Op Fl L | Fl -max-length Ar period
.Op Fl b | Fl -min-bitrate Ar bitrate
.Op Fl B | Fl -max-bitrate Ar bitrate
.Op Fl x | Fl -expression Ar regexp
.Op Fl X | Fl -exclude-expression Ar regexp
.Op Fl d | Fl -directory Ar directory
.Op Fl P | Fl -processes Ar count
.Op Fl v | Fl -invert
.Op Fl I | Fl -no-ignorecase
.Op Fl 0 | Fl -print0
.Pp
.Nm
.Brq Fl h | Fl -help
.
.
.Sh DESCRIPTION
.
The
.Nm
utility reads sequentially a list of ogg/vorbis files from standard input  and
filters this list using various criteria defined via command line options. All
ogg/vorbis files matching these criteria are written to standard output.
.Pp
All specified criteria are combined using logical AND.
.Pp
These are the available command line options:
.
.Bl -tag
.It Fl l | Fl -min-length Ar period
Matches every ogg/vorbis file with a play time longer than the specified 
.Ar period .
.Ar period 
may be expressed as seconds or in minutes:seconds syntax. 
.
.It Fl L | Fl -max-length Ar period
Matches every ogg/vorbis file with a play time shorter than the specified 
.Ar period .
.Ar period 
may be expressed as seconds or in minutes:seconds syntax. 
.
.It Fl b | Fl -min-bitrate Ar bitrate
Matches every ogg/vorbis file with a nominal bitrate higher than
.Ar bitrate
kbps
.
.It Fl B | Fl -max-bitrate Ar bitrate
Matches every ogg/vorbis file with a nominal bitrate lower than
.Ar bitrate
kbps
.
.It Fl x | Fl -expression Ar regexp
Matches every ogg/vorbis file containing at least one vorbiscomment matching
the extended regular expression
.Ar regexp .
The regular expression matching is case-insensitive unless the 
.Fl I
flag is given. See 
.Xr re_format 7
for a complete discussion of the extended regular expression format. This option
may be specified multiple times.
.
.It Fl X | Fl -exclude-expression Ar regexp
This is the inversion of the
.Fl x
flag. It excludes any ogg/vorbis files containing at least one vorbiscomment
matching the extended regular expression
.Ar regexp . 
This option may be specified multiple times.
.
.It Fl d | Fl -directory Ar directory
Prepends every line read from standard input with 
.Ar directory
if the first character of the line is not a slash.
.
.It Fl v | Fl -invert
Invert the result set - return all ogg/vorbis files not matching the specified
criteria.
.
.It Fl P | Fl -processes Ar count
Instructs
.Nm
to spawn 
.Ar count
worker processes. This may speed up the processing on multi processor machines.
A good choice is to set 
.Ar count
equal to the amount of processor cores available. By default 
.Nm
runs as a single process.
.
.It Fl I | Fl -no-ignorecase
This flag makes all expressions provided via
.Fl x
and
.Fl X
case sensitive.
.
.It Fl 0 | Fl -print0
This flag emits the matching filenames followed by an ASCII NUL character
instead of a newline character. You may use this to use
.Xr xargs 1
in conjunction with
.Nm .
.
.It Fl h | Fl -help
Print the synopsis of 
.Nm
and exit. This overrides any other options.
.
.El
.
.
.Sh ENVIRONMENT
.
.Nm
peruses the
.Ev LANG
and
.Ev LC_ALL
environment variables to determine the systems locale. This is mandatory for
expression matching.
.
.Sh EXAMPLES
.
To get a list of all your ogg/vorbis files tagged with genre 
.Dq Thrash Metal
use the following command line:
.Pp
.Dl find /my/music -type f -name '*.ogg' | oggfilter -x '^genre=thrash metal$'
.Pp
To filter a list of ogg/vorbis files for files not tagged as
.Dq Neo Folk
or
.Dq Power Metal
you may use:
.Pp
.Dl oggfilter -v -E -x '^genre=(neo folk|power metal)$' < playlist.m3u
.Pp
To get a list of all your ogg/vorbis files tagged with genre 
.Dq Thrash Metal
and containing
.Dq death
in it's title use the following command line:
.Pp
.Dl oggfilter -x '^genre=thrash metal$' -x '^title=.*death.*' < playlist.m3u
.Pp
To get a list of all your ogg/vorbis files tagged with genre 
.Dq Thrash Metal
and not containing
.Dq death
in it's title use the following command line:
.Pp
.Dl oggfilter -x '^genre=thrash metal$' -X '^title=.*death.*' < playlist.m3u
.Pp
To get a list of all ogg/vorbis files with a maximum playtime of 5 minutes and
a minimum playtime of 3 minutes you may use:
.Pp
.Dl oggfilter -l 180 -L 5:00 < playlist.m3u
.Pp
To get a list of ogg/vorbis files encoded with a minimal nominal bitrate of
120 kbps use:
.Pp
.Dl oggfilter -b 120 < playlist.m3u
.Pp
If you are piping from a playlist containing relative paths you can tell
.Nm
to prepend a base path to the read ogg/vorbis files:
.Pp
.Dl oggfilter -d /my/music -x '^genre=.*metal$' < relative.m3u
.
.
.Sh PORTABILITY
.
.Nm
was written in ANSI/C so it should compile on every unix-like platform
supporting an ANSI/C compiler and libvorbis. But this was written on a 
.Fx
system and has not been tested on other platforms. 
.Pp
The supplied Makefile is a BSD-style Makefile and uses the 
.Fx
infrastructure. This is not portable, but as
.Nm
is pretty simple there should be no problem to write a simple Makefile for
other platforms.
.
.
.Sh DIAGNOSTICS
.
.Ex -std
.
.
.Sh SEE ALSO
.
.Xr re_format 7 , 
.Xr vorbiscomment 1 ,
.Xr ogginfo 1 ,
.Xr setlocale 3 ,
.Xr environ 7 ,
.Xr xargs 1
.
.
.Sh AUTHORS
.
.An "Tobias Rehbein" Aq tobias.rehbein@web.de
.
.
.Sh BUGS
.
.Nm
will only work correctly in locales using 8-bit charsets or multi byte charsets
backwards compatible with ASCII, most notably UTF-8.
.Pp
Regular expression matching will only work if the vorbis comments are
representable in the locale charset.
.Pp
Expect some rough edges as this was my first take on a C program.
.Pp
If you find a bug drop me a mail or file an issue at 
.Dl http://github.com/blabber/oggfilter/issues
.
.


oggfilter.c
/*-
 * "THE BEER-WARE LICENSE" (Revision 42):
 * <tobias.rehbein@web.de> wrote this file. As long as you retain this notice
 * you can do whatever you want with this stuff. If we meet some day, and you
 * think this stuff is worth it, you can buy me a beer in return.
 *                                                              Tobias Rehbein
 */

#include <assert.h>
#include <err.h>
#include <limits.h>
#include <locale.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <sys/types.h>
#include <sys/wait.h>

#include "checks.h"
#include "list.h"
#include "options.h"

static void			 fork_you(struct opt_options *opts, struct chk_context *ctx);
static void			 free_conditions(struct chk_conditions *cond);
static struct chk_conditions 	*get_conditions(struct opt_options *opts);
static void			 process_loop(struct opt_options *opts, struct chk_context *ctx, bool doflush);
static void			 remove_trailing_slashes(char *path);
static int			 use_pipe(int *p);
static void			 wait_for_childs(void);

int
main(int argc, char **argv)
{

	if (setlocale(LC_ALL, "") == NULL)
		errx(EXIT_FAILURE, "setlocale LC_ALL");

	/* setup environment */
	struct opt_options *opts = opt_get_options(argc, argv);
	if (opts == NULL)
		err(EXIT_FAILURE, "opt_get_options");
	struct chk_conditions *cond = get_conditions(opts);
	if (cond == NULL)
		err(EXIT_FAILURE, "get_conditions");
	struct chk_context *ctx = chk_context_open(cond);
	if (ctx == NULL)
		err(EXIT_FAILURE, "chk_context_open");

	/* enter main loop or fork away */
	if (opts->processes <= 1)
		process_loop(opts, ctx, false);
	else
		fork_you(opts, ctx);

	/* free all resources */
	chk_context_close(ctx);
	free_conditions(cond);
	opt_free_options(opts);

	return (0);
}

static struct chk_conditions *
get_conditions(struct opt_options *opts)
{
	assert(opts != NULL);

	struct chk_conditions *cond = malloc(sizeof(*cond));
	if (cond == NULL)
		return (NULL);

	chk_init_conditions(cond);

	cond->min_length = opts->min_length;
	cond->max_length = opts->max_length;
	cond->min_bitrate = opts->min_bitrate;
	cond->max_bitrate = opts->max_bitrate;
	cond->noignorecase = opts->noignorecase;

	for (struct element *oe = opts->expressionlist; oe != NULL; oe = oe->next) {
		struct chk_expression *cx = malloc(sizeof(*cx));
		if (cx == NULL)
			err(EXIT_FAILURE, "malloc chk_expression");

		struct opt_expression *ox = oe->payload;
		cx->expression = ox->expression;
		cx->invert = ox->invert;

		struct element *ce;
		if ((ce = create_element(cx)) == NULL)
			err(EXIT_FAILURE, "create_element");

		cond->regexlist = prepend_element(ce, cond->regexlist);
	}

	return (cond);
}

static void
free_conditions(struct chk_conditions *cond)
{
	assert(cond != NULL);

	for (struct element *e = cond->regexlist; e != NULL; e = destroy_element(e))
		free(e->payload);

	free(cond);
}

static void
process_loop(struct opt_options *opts, struct chk_context *ctx, bool doflush)
{
	assert(opts != NULL);
	assert(ctx != NULL);

	char pathread[LINE_MAX];
	while (fgets(pathread, LINE_MAX, stdin) != NULL) {
		char *newline = NULL;
		if ((newline = strchr(pathread, '\n')) != NULL)
			newline[0] = '\0';

		char *path = malloc(LINE_MAX);
		if (path == NULL)
			err(EXIT_FAILURE, "malloc path");

		char *pathprefix = NULL;
		if (opts->pathprefix != NULL) {
			pathprefix = strdup(opts->pathprefix);
			if (pathprefix == NULL)
				err(EXIT_FAILURE, "strdup pathprefix");
			remove_trailing_slashes(pathprefix);
		}
		if (pathprefix != NULL && pathread[0] != '/')
			snprintf(path, LINE_MAX, "%s/%s", pathprefix, pathread);
		else {
			strncpy(path, pathread, LINE_MAX - 2);
			path[LINE_MAX - 1] = '\0';
		}
		free(pathprefix);

		bool check_result = chk_check_file(path, ctx);
		if (check_result ^ opts->invert) {
			fputs(path, stdout);
			if (opts->print0)
				putc('\0', stdout);
			else
				putc('\n', stdout);

			if (doflush)
				fflush(stdout);
		}
		free(path);
	}
	if (ferror(stdin))
		err(EXIT_FAILURE, "fgets stdin");
}

static void
remove_trailing_slashes(char *path)
{
	assert(path != NULL);

	char *n = strchr(path, (int)'\0');
	assert(n != NULL);
	while (*n != path[0] && *(--n) == '/')
		*n = '\0';
}

static void
fork_you(struct opt_options *opts, struct chk_context *ctx)
{
	assert(opts != NULL);
	assert(ctx != NULL);

	int *fds = malloc(opts->processes * sizeof(int));
	if (fds == NULL)
		err(EXIT_FAILURE, "malloc fds");

	for (int i = 0; i < opts->processes; i++) {
		int p[2];
		if (pipe(p) == -1)
			err(EXIT_FAILURE, "pipe %d", i);

		pid_t pid = fork();
		switch (pid) {
		case -1:
			err(EXIT_FAILURE, "fork %d", i);
			break;
		case 0:
			/* close all write ends opened until now */
			for (int j = 0; j < i; j++)
				if (close(fds[j]) == -1)
					warn("close %d, %d", i, j);
			free(fds);

			if (use_pipe(p) == -1)
				errx(EXIT_FAILURE, "use_pipe %d", i);

			process_loop(opts, ctx, true);

			return;
		default:
			/*
			 * close the read end of pipe and store write end for
			 * later use
			 */
			if (close(p[0]) == -1)
				warn("close %d", i);
			fds[i] = p[1];
		}
	}

	char pathread[LINE_MAX];
	int fd = 0;
	while (fgets(pathread, LINE_MAX, stdin) != NULL)
		if (write(fds[fd++ % opts->processes], pathread, strlen(pathread)) == -1)
			warn("write");
	if (ferror(stdin))
		err(EXIT_FAILURE, "fgets stdin");

	for (int i = 0; i < opts->processes; i++)
		if (close(fds[i]) == -1)
			warn("close %d", i);

	wait_for_childs();

	free(fds);
}

static int
use_pipe(int *p)
{
	assert(p != NULL);

	/* close write end of pipe and stdin of process */
	if (close(p[1]) == -1)
		return (-1);
	if (fclose(stdin) == EOF)
		return (-1);

	/* use the read end of pipe as stdin */
	if ((dup2(p[0], STDIN_FILENO)) == -1)
		return (-1);
	if ((stdin = fdopen(STDIN_FILENO, "r")) == NULL)
		return (-1);

	return (0);
}

static void
wait_for_childs()
{
	pid_t pid;
	int sts;
	while ((pid = wait(&sts)) != -1)
		if (WIFEXITED(sts) && WEXITSTATUS(sts) != 0)
			warn("child process %d exited abnormally: %d", pid, sts);
}
